### PR TITLE
Feature/command line arguments

### PR DESCRIPTION
**Pull request by theMK2k**
Please review my attempt at introducing:
- simple parsing of command line arguments
- auto-run a game (together with core and sytem) - all of them provided as command line arguments to RALibretro
- see added docs in [README.md](https://github.com/theMK2k/RALibretro/tree/feature/command_line_arguments#command-line-arguments)

Issue: https://github.com/RetroAchievements/RALibretro/issues/66

### CODE REVIEW

File: src/Application.cpp [hunk 1]
```diff
@@ -2509,3 +2511,78 @@ void Application::toggleFastForwarding(unsigned extra)
...
+        break;
+      
+      default:
+        break;
+    }
+  }
+  return result;
+}
+
+bool Application::tryLoadContent(parsedArgs parsedArgs)
```
Comment (Jamiras) [inline src/Application.cpp:2566]: Recommend passing by const reference (`const parsedArgs&`)

File: src/Application.cpp [hunk 2]
```diff
@@ -2509,3 +2511,78 @@ void Application::toggleFastForwarding(unsigned extra)
...
+{
+  if (parsedArgs.core.length() == 0 || parsedArgs.system == 0 || parsedArgs.game.length() == 0) {
+    return true;  // nothing to do
+  }
+
+  if (!util::exists(parsedArgs.game)) {
+    _logger.error("file not found provided in 'game' command line argument '%s'", parsedArgs.game.c_str());
```
Comment (Jamiras) [inline src/Application.cpp:2573]: This should also be shown as a messagebox so the user understands what went wrong (see how `Application::loadCore` handles "Could not load "...)

File: src/Application.h [hunk 3]
```diff
@@ -40,12 +40,22 @@ along with RALibretro.  If not, see <http://www.gnu.org/licenses/>.
 #include "Memory.h"
 #include "States.h"
 
+/* Parsed command line arguments */
+struct parsedArgs
```
Comment (Jamiras) [inline src/Application.h:44]: Please use PascalCase for type names

File: src/main.cpp [hunk 4]
```diff
@@ -74,6 +74,13 @@ int main(int argc, char* argv[])
 
   bool ok = app.init("RALibRetro", 640, 480);
 
+  parsedArgs parsedArgs = app.parseArgs(argc, argv);
+  app.logger().info("[ARGS] core: %s", parsedArgs.core.c_str());
+  app.logger().info("[ARGS] system: %i", parsedArgs.system);
+  app.logger().info("[ARGS] game: %s", parsedArgs.game.c_str());
+
+  ok &= app.tryLoadContent(parsedArgs);
```
Comment (Jamiras) [inline src/main.cpp:82]: suggest wrapping this new code in an
```
if (ok && argc > 1) {
```
block to avoid logging
```
[INFO] [ARGS] core: 
[INFO] [ARGS] system: 0
[INFO] [ARGS] game: 
```
Also, recommend just using "[APP]" as the category.

Comment (Jamiras) [reply src/main.cpp:82]: VisualStudio doesn't like something about the way this is interacting with the `__try` below:
```
1>e:\source\retroachievements\ralibretro\src\main.cpp(71): error C2712: Cannot use __try in functions that require object unwinding
```
You'll probably have to move it into a function.

File: src/Application.cpp [hunk 5]
```diff
@@ -2509,3 +2511,78 @@ void Application::toggleFastForwarding(unsigned extra)
...
+    return false;
+  }
+  
+  RA_AttemptLogin(true); // attempt login with blocking, so that we go on exactly after the login was successful
```
Comment (Jamiras) [inline src/Application.cpp:2577]: `app.init()` will have already kicked off an asynchronous login by this point. You might want to reorder things so you can avoid the asynchronous login in this code path.

File: src/Application.cpp [hunk 6]
```diff
@@ -52,6 +52,8 @@ along with RALibretro.  If not, see <http://www.gnu.org/licenses/>.
 
 #define TAG "[APP] "
 
+#include <thread>
```
Comment (Jamiras) [inline src/Application.cpp:55]: Why is this necessary?

File: README.md [hunk 7]
```diff
@@ -48,3 +48,15 @@ $ make
...
+
+## Command Line Arguments
+
+Argument|Description
+-|-
+-c [--core]|the core's name, e.g. `--core picodrive_libretro`
+-s [--system]|the system id, see ConsoleID in [RAInterface/RA_Consoles.h](https://github.com/RetroAchievements/RAInterface/blob/master/RA_Consoles.h), e.g. `--system 1`
+-g [--game]|full path to the game's file, e.g. `--game "C:\ROMS\GEN\Demons Of Asteborg Demo.gen"`
+
+In order to successfully run a game on startup, provide the core, system and game, e.g.:
```
Comment (Jamiras) [inline README.md:60]: I'd leave out the 'successfully'. Just because you provide all three parameters doesn't mean it will succeed. You could have a typo, or put in the wrong system id.

File: src/main.cpp (hunk 4 above)
Comment (theMK2k) [reply src/main.cpp:82]: @Jamiras thanks for your comments, I'll try to solve them all the next days. This one however, I can't properly test, because I don't have VisualStudio - the MinGW build runs through.

File: src/Application.h (hunk 3 above)
Comment (theMK2k) [reply src/Application.h]: ✔️ https://github.com/theMK2k/RALibretro/commit/a1208ef0972cce680ba810ea48f70d4c9f1eeff5

File: src/Application.cpp (hunk 1 above)
Comment (theMK2k) [reply src/Application.cpp]: ✔️ https://github.com/theMK2k/RALibretro/commit/a34042c805d5b3ec3a83ed4c7651b09e28592dba

File: src/Application.cpp (hunk 6 above)
Comment (theMK2k) [reply src/Application.cpp]: it isn't (anymore)

✔️ https://github.com/theMK2k/RALibretro/commit/86769071666acd9383bcfb512bb414e374548a10

File: src/Application.cpp (hunk 2 above)
Comment (theMK2k) [reply src/Application.cpp]: ✔️ https://github.com/theMK2k/RALibretro/commit/a2b0d1e379d343b35a122f9bd9fd19e31741c485

File: src/main.cpp (hunk 4 above)
Comment (theMK2k) [reply src/main.cpp]: ✔️ (check if any cli args are provided, log as [APP]) https://github.com/theMK2k/RALibretro/commit/160681e1219745d515917547a2e17493dff4d350

File: README.md (hunk 7 above)
Comment (theMK2k) [reply README.md]: ✔️ https://github.com/theMK2k/RALibretro/commit/92028804fa120165b0da8bce7bc953d5c6b21204

**Comment by Jamiras** [comment]
I made some changes to address the `Cannot use __try in functions that require object unwinding` error. Primarily merging the `parseArgs` and `tryLoadContent` functions into a single function so the `ParseArgs` object didn't have to be allocated in `main`.

Also addressed my comment about the double login and added some code to validate the specified system is valid for the specified core.

I hope you don't mind. Let me know if you have any questions or concerns.

**Comment by theMK2k** [comment]
@Jamiras thanks for helping out for the remaining issues. I don't mind at all and your solution for the "do not attempt a double login" is interesting by just preventing the original non-blocking login.

I tested with working and non-working cli args examples after building locally. Looks good :)

**Review by Jamiras — APPROVED** [review]
